client/state/sites/vouchers/reducer.js

```diff
@@ -26,14 +30,39 @@ import { itemsSchema } from './schema';
  * @return {Object} updated state
  */
 export const items = ( state = {}, action ) => {
-	const { siteId } = action;
-	switch ( action.type ) {
+	const {
+		siteId,
+		type,
+		voucher,
+		vouchers,
+		serviceType
+	} = action;
+
+	switch ( type ) {
+		case SITE_VOUCHERS_ASSIGN_RECEIVE:
+			const serviceVouchers = state[ siteId ]
+				? ( state[ siteId ][ serviceType ] || [] )
+				: [];
+
+			return Object.assign(
+				{},
+				state,
+				{
+					[ siteId ]: Object.assign(
+						{},
+						{ [ serviceType ]: serviceVouchers.concat( voucher ) }
+					)
```

Comment:
This assignment-clone is redundant here and doesn't provide us anything. This is mainly a comment-bump from [my earlier comment](https://github.com/Automattic/wp-calypso/pull/5621#discussion_r64846952). We don't risk anything by taking out the second `Object.assign()`

```js
return Object.assign( {}, state, {
	[ siteId ]: serviceWorkers.concat( voucher )
} );
```